### PR TITLE
Update SetTrustLineFlagsOpFrame / AllowTrustOpFrame to handle auth revocation for CAP-0038

### LINKED ISSUE #3108: Update SetTrustLineFlagsOpFrame / AllowTrustOpFrame for CAP-38
- Implement new `LedgerTxn` and `LedgerTxnRoot` queries to load `LiquidityPoolEntry` by account and asset (#3140)
- Implement a utility function to compute withdrawal amounts (also needed to handle normal withdrawals) (#3147)
- Update `removeOffersByAccountAndAsset` to also remove `LiquidityPoolEntry` (#3162)

### PR DESCRIPTION
**Pull request by sisuresh**
# Description

Resolves #3108

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/TransactionUtils.cpp
Comment (jonjove) [inline src/transactions/TransactionUtils.cpp:1243]: You might consider refactoring this block and the next into a `makeRevokeID` function.

File: src/transactions/TransactionUtils.cpp
Comment (jonjove) [inline src/transactions/TransactionUtils.cpp:1245]: This doesn't match our normal naming convention. I think we should fix the CAP so that it's `revokeID` not `revokeId` (see `AccountID`, `OperationID`, `BalanceID`, etc for reference).

File: src/transactions/TransactionUtils.cpp
Comment (jonjove) [inline src/transactions/TransactionUtils.cpp:1260]: You might consider refactoring this and the next two blocks into a `makeRevokeClaimants` function (or something similar).

File: src/transactions/TransactionUtils.cpp
Comment (jonjove) [inline src/transactions/TransactionUtils.cpp:1276]: Nitpick: can we use `loadTrustLine` here?

File: src/transactions/TransactionUtils.cpp
Comment (jonjove) [inline src/transactions/TransactionUtils.cpp:1227]: It would be really great if we could erase `poolShareTrustLine` at the same time as we release the reserves. As part of #3169, I had to deal with some cases like `ManageOfferOpFrameBase` where we erased at a different place and it is really confusing. Looking at the rest of this function, it seems like this should be feasible.

File: src/transactions/TransactionUtils.cpp
Comment (jonjove) [inline src/transactions/TransactionUtils.cpp:1233]: Nitpick: no reason to take `txSeqNum` and `opIndex` as references, they are integers.

File: src/transactions/TransactionUtils.cpp
Comment (jonjove) [reply src/transactions/TransactionUtils.cpp:1260]: Nitpick: also for this use case, I don't see a lot of reason to copy the `ClaimPredicate` in. You could have just written `c.v0().predicate.type(CLAIM_PREDICATE_UNCONDITIONAL)` on line 1265.

File: src/transactions/TransactionUtils.cpp
Comment (jonjove) [inline src/transactions/TransactionUtils.cpp:1215]: This would also probably look a lot nicer if written as a little function.

**Review by jonjove — CHANGES_REQUESTED** [review]
My main comment is that `removeOffersByAccountAndAsset` is too hard to validate/verify right now. I've added many comments about how to simplify that so it can be reviewed more easily. Once that's done, I'll revisit that function to check it for correctness then review the tests.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1276]: Not without making some other changes. We need to access the extension to modify `liquidityPoolUseCount`.

Related to this though, there's a bug here. If `accountID` is the issuer of `assetInPool`, then this will throw. This is possible because the other asset (not the one being revoked) can be issued by `accountID`. This is easy to fix, but the interesting question to ask here is should we bother creating the claimable balance in this case? I don't think we should since the claimable balance will just be "burned" when it's claimed. Either way, I'll update the CAP to clarify this detail once we decide what to do.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Updated to `removeOffersAndPoolShareTrustLines`.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: done

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Good idea. Done.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Done

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Done

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Fixed

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Updated to use `makeUnconditionalClaimant`.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: I updated the implementation to not create the claimable balance if the claimant is the issuer. I'll also update the CAP and post the link to the PR here.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Done

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Fixed

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Added a `getTrustLineBacker` function for this.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: https://github.com/stellar/stellar-protocol/pull/1042

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [inline src/transactions/TransactionUtils.cpp:1344]: There's a bug here. We don't always reach this line (ex. when pool share trust line balance is 0).

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: I should also prefetch the asset trustlines.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Fixed and added a test.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: Done, but it required two separate prefetches

File: src/transactions/TransactionUtils.cpp
Comment (MonsieurNicolas) [inline src/transactions/TransactionUtils.cpp:1393]: (CAP question I guess)

Is that true? If there is a sponsorship sandwich, we're releasing an entry from `cbSponsoringAccID` but adding one to the sponsor

File: src/transactions/SetTrustLineFlagsOpFrame.cpp
Comment (MonsieurNicolas) [inline src/transactions/SetTrustLineFlagsOpFrame.cpp:81]: nit: missed comment on L55

File: src/transactions/TransactionUtils.cpp
Comment (MonsieurNicolas) [inline src/transactions/TransactionUtils.cpp:1379]: this is a bit sad that it's basically copy/pasted from `CreateClaimableBalanceOpFrame`

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
I tried and failed to find the bug that you mentioned. The code is quite complicated, so I would not be surprised if there was something related to sponsorships.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1220]: We need to load the liquidity pools to get the assets for the asset trustlines, so those are prefetched first.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1393]: You're right - If the sponsoring account has `numSponsoring + numSubentries == UINT32_MAX` (this limit is in the CAP but not master yet), then this will throw. We should instead set `opTOO_MANY_SPONSORING` like we do everywhere else.

I'll update the CAP as well. The issuer just needs to submit the operation outside of a sponsorship sandwich for this to succeed.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1477]: For the usage of this method during auth revocation, it would deactivate the entries loaded with `loadPoolShareTrustLinesByAccountAndAsset`. This is a good point though.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1393]: https://github.com/stellar/stellar-protocol/pull/1044

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1288]: If `ltxInner` is passed into `removeOffersByAccountAndAsset`, then `ltxInner.commit()` will need to be called whenever this function returns early (eg. if `poolShareTrustLines` is empty). I thought it would be safer this way in case another termination condition is added without the `commit`.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1379]: I'm not sure if it's worth unifying the two blocks. They are different enough where I think the code would just get more complex.

File: src/xdr/Stellar-ledger-entries.x
Comment (sisuresh) [reply src/xdr/Stellar-ledger-entries.x:402]: Yeah I think removing it makes sense.

File: src/transactions/TransactionUtils.cpp
Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp]: removed

File: src/transactions/SetTrustLineFlagsOpFrame.cpp
Comment (sisuresh) [reply src/transactions/SetTrustLineFlagsOpFrame.cpp]: fixed

File: src/xdr/Stellar-transaction.x
Comment (sisuresh) [reply src/xdr/Stellar-transaction.x:539]: Updated

Comment (sisuresh) [reply src/xdr/Stellar-transaction.x:539]: https://github.com/stellar/stellar-protocol/pull/1045

**Review by jonjove — CHANGES_REQUESTED** [review]

File: src/test/TxTests.cpp
Comment (jonjove) [reply src/test/TxTests.cpp:413]: Let's make this a `static` function in SetTrustLineFlagsTests.cpp then. The fact that people can use it with the wrong parameter makes it kind of dangerous.

Comment (Diego77777777777) [reply src/test/TxTests.cpp:413]: ****

Comment (sisuresh) [reply src/test/TxTests.cpp]: Done

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (jonjove) [inline src/transactions/test/SetTrustLineFlagsTests.cpp:662]: What about similar checks for the non-native asset?

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (jonjove) [inline src/transactions/test/SetTrustLineFlagsTests.cpp:715]: Probably worth checking the case where only one balance is 0 due to rounding.

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (jonjove) [inline src/transactions/test/SetTrustLineFlagsTests.cpp:989]: I like this. Tests like this are essentially what I had in mind in my long comment above.

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (jonjove) [inline src/transactions/test/SetTrustLineFlagsTests.cpp:950]: Probably don't want to commit on failure.

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (sisuresh) [reply src/transactions/test/SetTrustLineFlagsTests.cpp]: Fixed

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (sisuresh) [reply src/transactions/test/SetTrustLineFlagsTests.cpp]: This should be fixed with the refactor.

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (sisuresh) [reply src/transactions/test/SetTrustLineFlagsTests.cpp]: This should be fixed with the refactor.

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (sisuresh) [reply src/transactions/test/SetTrustLineFlagsTests.cpp]: Fixed

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (sisuresh) [reply src/transactions/test/SetTrustLineFlagsTests.cpp]: Fixed

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (sisuresh) [reply src/transactions/test/SetTrustLineFlagsTests.cpp:495]: I refactored most of the `revoke from pool` test. It's not exactly what you suggested, but it's close and I think the tests are much more thorough and easier to understand. Let me know what you think.

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (sisuresh) [reply src/transactions/test/SetTrustLineFlagsTests.cpp:636]: That test is already on our todo list of CAP-0038 tests. I think we can implement it in a subsequent PR to avoid holding this PR up. Let me know if you disagree.

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (jonjove) [reply src/transactions/test/SetTrustLineFlagsTests.cpp:495]: Awesome, I was able to review this in single digit minutes now. Really nicely done.

File: src/transactions/test/SetTrustLineFlagsTests.cpp
Comment (jonjove) [reply src/transactions/test/SetTrustLineFlagsTests.cpp:636]: Yes that's fine.

**Review by MonsieurNicolas — APPROVED** [review]
LGTM!

File: src/transactions/SponsorshipUtils.cpp
Comment (MonsieurNicolas) [inline src/transactions/SponsorshipUtils.cpp:186]: note that we also 2 other implementations of this function --- and I just opened https://github.com/stellar/stellar-core/pull/3192 because returning uint32 is a footgun. So you probably want to move to int32 and use that util in test code as well after that other PR gets merged?

Comment (sisuresh) [reply src/transactions/SponsorshipUtils.cpp]: Done. I moved the test implementation into `TestUtils` to keep it separate. After all, these tests are testing `computeMultiplier` as well.

**Comment by MonsieurNicolas** [comment]
r+ dbfed84428877bd216d298410dcadc01d0191350

File: src/transactions/TransactionUtils.cpp
Comment (bartekn) [inline src/transactions/TransactionUtils.cpp:1262]: I decided to ask here first before creating a new issue. Should it set the type to `CLAIMABLE_BALANCE_ID_TYPE_FROM_POOL_REVOKE` instead of `CLAIMABLE_BALANCE_ID_TYPE_V0`?

Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1262]: `CLAIMABLE_BALANCE_ID_TYPE_FROM_POOL_REVOKE` was removed from the CAP. I would look over https://github.com/stellar/stellar-protocol/commits/master/core/cap-0038.md as there have been some minor changes recently.

Comment (bartekn) [reply src/transactions/TransactionUtils.cpp:1262]: It still says:
> The claimable balances will have a different `ClaimableBalanceIDType` to indicate their origin.

I will open a PR in stellar-protocol repo. Thanks!

Comment (sisuresh) [reply src/transactions/TransactionUtils.cpp:1262]: Ah good catch

Comment (bartekn) [reply src/transactions/TransactionUtils.cpp:1262]: https://github.com/stellar/stellar-protocol/pull/1048

**Review by deauna — COMMENTED** [review]
Send tokens to my federation address on Stellar. No memo required: SEP-0002*lobstr.co

**Pull request merged**
